examples/mlperf/rnnt/train.py

```diff
@@ -0,0 +1,367 @@
+    t = self.l1(inp).relu()
+    t = t.dropout(self.dropout)
+    return self.l2(t)
+
+class Encoder:
+  def __init__(self, input_size, hidden_size, pre_layers, post_layers, stack_time_factor, dropout):
+    self.pre_rnn = LSTM(input_size, hidden_size, pre_layers, dropout)
+    self.stack_time = StackTime(stack_time_factor)
+    self.post_rnn = LSTM(stack_time_factor * hidden_size, hidden_size, post_layers, dropout)
+    self.params = [*self.pre_rnn.params, *self.post_rnn.params]
+
+  def __call__(self, x:Tensor, x_lens):
+    x, _ = self.pre_rnn(x, None)
+    x, x_lens = self.stack_time(x, x_lens)
+    x, _ = self.post_rnn(x, None)
+    return x.transpose(0, 1), x_lens
+  
+class StackTime:
+  def __init__(self, factor):
+    self.factor = factor
+
+  def __call__(self, x, x_lens):
+    x = x.pad(((0, (-x.shape[0]) % self.factor), (0, 0), (0, 0)))
+    x = x.reshape(x.shape[0] // self.factor, x.shape[1], x.shape[2] * self.factor)
+    return x, x_lens / self.factor if x_lens is not None else None
+
+class Prediction:
+  def __init__(self, vocab_size, hidden_size, layers, dropout):
+    self.hidden_size = hidden_size
+
+    self.emb = Embedding(vocab_size - 1, hidden_size)
+    self.rnn = LSTM(hidden_size, hidden_size, layers, dropout)
+
+    self.params = [self.emb.weight,*self.rnn.params]
+
+  def __call__(self, x, hc, m):
+    emb = self.emb(x) * m
+    x_, hc = self.rnn(emb.transpose(0, 1), hc)
+    return x_.transpose(0, 1), hc
+
+# %% [markdown]
+# # training
+
+# %%
+def calc_loss(enc,distribution):
+
+
+    T,U = distribution.shape[2],distribution.shape[1]
+    assert len(labels) == U-1
+    assert enc.shape[1] == T, f"{enc.shape}[1] != {T}"
+
+    alpha = np.zeros((T,U))
+    alpha[0,0] = 1.
+
+    for t in range(T):
+        for u in range(U):
+            if t < T-1: 
+                alpha[t+1,u] = alpha[t,u] * distribution[0,u,t,-1]
+            if u < U-1: 
+                label = int((labels)[u])
+                alpha[t,u+1] += alpha[t,u] * distribution[0,u,t,label]
+
+    P = alpha[-1,-1] * distribution[0,-1,-1,-1]
+    Loss = -np.log(P)
+    Loss_grad = 1
+
+    alpha_grad = np.zeros((T,U))
+    alpha_grad[-1,-1] = -1/alpha[-1,-1]
+    distribution_grad = np.zeros_like(distribution,dtype=np.float64)
+    distribution_grad[0,-1,-1,-1] = -1/distribution[0,-1,-1,-1]
+
+    for t in reversed(range(T)):
+        for u in reversed(range(U)):
+            if t < T-1:
+                alpha_grad[t,u] += alpha_grad[t+1,u] * distribution[0,u,t,-1]
+                distribution_grad[0,u,t,-1] += alpha_grad[t+1,u] * alpha[t,u]
+            if u < U-1:
+                label = int((labels)[u])
+
+                alpha_grad[t,u] += alpha_grad[t,u+1] * distribution[0,u,t,label]
+                distribution_grad[0,u,t,label] += alpha_grad[t,u+1] * alpha[t,u]
+
+    # plt.imshow(-distribution_grad[0,:,:,:].sum(-1))
+    # plt.show()
+    return Loss,distribution_grad
+
+# %%
+rnnt= RNNT()
+
+# %%
+opt = LAMB(rnnt.params)
+def train_step():
```

Comment:
This is where `@TinyJit` should be used. You feed `model`, `optimizer`, (`schedular` if there is one), `X` and `Y`

Reply:
I just copied this code from https://github.com/tinygrad/tinygrad/blob/master/extra/datasets/librispeech.py can translate to tensors but this is all just a couple of seconds for the whole dataset.

Reply:
okay will do.